src/Entity/SmsMessage.php

```diff
@@ -134,25 +135,73 @@ public function removeOption($name) {
    * {@inheritdoc}
    */
   public function getResult() {
-    return $this->get('result')->entity;
+    $results = $this->entityTypeManager()
+      ->getStorage('sms_result')
+      ->loadByProperties(['parent' => $this->id()]);
+    return $results ? reset($results) : NULL;
   }
 
   /**
    * {@inheritdoc}
    */
   public function setResult(StdMessageResultInterface $result = NULL) {
     if ($result === NULL) {
-      $this->set('result', NULL);
+      // Find any pre-existing report and remove the parent-child relationship.
+      if ($previous_result = $this->getResult()) {
```

Comment:
I decided to make the result immutable. So `setResult()` will only work if an existing result wasn't saved before. An exception will be thrown if attempt is made to update an existing result.